### ISSUE TITLE
How to detect the horizontal text only?

### ISSUE THREAD

**Issue by zqm840527** [argman/EAST#193]
Is it possible to slightly modify the code and make EAST just detect the horizontal text only?

I am aware that CTPN only detects the horizontal text, but from my testing result, I found EAST is faster, and the accuracy also slightly better, so I prefer to stick on EAST.

So, could anyone help me to make it happen? Thanks.

**Comment by ShiyangZhang** [comment]
@zqm840527  
two methohs.
It is possible and quit easy.
you can  filter rotated labels when training your model and also when you get angles of the rotated boxes. 
see https://github.com/argman/EAST/blob/ab97939783901b7e22ff55e151964e159d1627b9/eval.py#L69

**Comment by zqm840527** [comment]
@ShiyangZhang 
从您的名字我猜您是华人，我的英文不好，所以直接打中文了。

首先感谢您的指点，我也看明白了是要过滤掉和旋转相关的标签，但我对编程的了解实在是太有限，没搞明白具体要怎么实施。

我的目的是运行run_demo_server.py的时候只探测水平文字，忽略掉所有有旋转角度的。我直接使用训练好的模型，不再进行新的训练。

如果不占用您太多时间的话，可否具体解释一下要注释掉哪些行，修改哪些参数？或者能否麻烦您直接分享一个修改好的文件供我替换？

如果需要邮件沟通的话，我的邮箱是zqm840527@hotmail.com。

万分感谢。

**Comment by ShiyangZhang** [comment]
> @ShiyangZhang
> 从您的名字我猜您是华人，我的英文不好，所以直接打中文了。
> 
> 首先感谢您的指点，我也看明白了是要过滤掉和旋转相关的标签，但我对编程的了解实在是太有限，没搞明白具体要怎么实施。
> 
> 我的目的是运行run_demo_server.py的时候只探测水平文字，忽略掉所有有旋转角度的。我直接使用训练好的模型，不再进行新的训练。
> 
> 如果不占用您太多时间的话，可否具体解释一下要注释掉哪些行，修改哪些参数？或者能否麻烦您直接分享一个修改好的文件供我替换？抱歉能力有限，只能做一回伸手党。
> 
> 如果需要私下邮件沟通的话，我的邮箱是[zqm840527@hotmail.com](mailto:zqm840527@hotmail.com)。
> 
> 万分感谢。

sure. but maybe latter. i will change some code below
https://github.com/argman/EAST/blob/ab97939783901b7e22ff55e151964e159d1627b9/run_demo_server.py#L123

**Comment by zqm840527** [comment]
@ShiyangZhang

it seems that is a little complicated than we expect...but thanks anyway.

**Comment by zqm840527** [comment]
With Mr.@ShiyangZhang's help, I am able to get what I asked.
----------------------------------------------------
#!/usr/bin/env python3

import os

import time
import datetime
import cv2
import numpy as np
import uuid
import json

import functools
import logging
import collections
import math

logger = logging.getLogger(__name__)
logger.setLevel(logging.INFO)


@functools.lru_cache(maxsize=1)
def get_host_info():
    ret = {}
    with open('/proc/cpuinfo') as f:
        ret['cpuinfo'] = f.read()

    with open('/proc/meminfo') as f:
        ret['meminfo'] = f.read()

    with open('/proc/loadavg') as f:
        ret['loadavg'] = f.read()

    return ret


@functools.lru_cache(maxsize=100)
def get_predictor(checkpoint_path):
    logger.info('loading model')
    import tensorflow as tf
    import model
    from icdar import restore_rectangle
    import lanms
    from eval import resize_image, sort_poly, detect

    input_images = tf.placeholder(tf.float32, shape=[None, None, None, 3], name='input_images')
    global_step = tf.get_variable('global_step', [], initializer=tf.constant_initializer(0), trainable=False)

    f_score, f_geometry = model.model(input_images, is_training=False)

    variable_averages = tf.train.ExponentialMovingAverage(0.997, global_step)
    saver = tf.train.Saver(variable_averages.variables_to_restore())

    sess = tf.Session(config=tf.ConfigProto(allow_soft_placement=True))

    ckpt_state = tf.train.get_checkpoint_state(checkpoint_path)
    model_path = os.path.join(checkpoint_path, os.path.basename(ckpt_state.model_checkpoint_path))
    logger.info('Restore from {}'.format(model_path))
    saver.restore(sess, model_path)

    def predictor(img):
        """
        :return: {
            'text_lines': [
                {
                    'score': ,
                    'x0': ,
                    'y0': ,
                    'x1': ,
                    ...
                    'y3': ,
                }
            ],
            'rtparams': {  # runtime parameters
                'image_size': ,
                'working_size': ,
            },
            'timing': {
                'net': ,
                'restore': ,
                'nms': ,
                'cpuinfo': ,
                'meminfo': ,
                'uptime': ,
            }
        }
        """
        start_time = time.time()
        rtparams = collections.OrderedDict()
        rtparams['start_time'] = datetime.datetime.now().isoformat()
        rtparams['image_size'] = '{}x{}'.format(img.shape[1], img.shape[0])
        timer = collections.OrderedDict([
            ('net', 0),
            ('restore', 0),
            ('nms', 0)
        ])

        im_resized, (ratio_h, ratio_w) = resize_image(img)
        rtparams['working_size'] = '{}x{}'.format(
            im_resized.shape[1], im_resized.shape[0])
        start = time.time()
        score, geometry = sess.run(
            [f_score, f_geometry],
            feed_dict={input_images: [im_resized[:,:,::-1]]})
        timer['net'] = time.time() - start

        boxes, timer = detect(score_map=score, geo_map=geometry, timer=timer)
        logger.info('net {:.0f}ms, restore {:.0f}ms, nms {:.0f}ms'.format(
            timer['net']*1000, timer['restore']*1000, timer['nms']*1000))

        if boxes is not None:
            scores = boxes[:,8].reshape(-1)
            boxes = boxes[:, :8].reshape((-1, 4, 2))
            boxes[:, :, 0] /= ratio_w
            boxes[:, :, 1] /= ratio_h

        duration = time.time() - start_time
        timer['overall'] = duration
        logger.info('[timing] {}'.format(duration))

        text_lines = []
		
        def get_slope(p):
            '''
            use this function after sort_poly
           
            '''
            import math
            #p = sort_poly(p)
           
            side_len_0 = math.sqrt((p[0, 0] - p[1, 0])**2 +(p[0, 1] - p[1, 1])**2)
            side_len_1 = math.sqrt((p[2, 0] - p[1, 0])**2 +(p[2, 1] - p[1, 1])**2)
            if side_len_0 > side_len_1:
                slope = (p[1, 1] - p[0, 1]) *1.0/(abs(p[1, 0] - p[0, 0]) + 1e-5)
            elif side_len_0 < side_len_1:
                slope = (p[2, 1] - p[1, 1]) *1.0/(abs(p[2, 0] - p[1, 0]) + 1e-5)
            else:
                #impossible
                print("unknow angle")
                slope = (p[2, 1] - p[1, 1]) *1.0/(abs(p[2, 0] - p[1, 0]) + 1e-5)
            #print(side_len_0, side_len_1,slope, p)
            return math.atan(slope)/math.pi*180
			
			
			
			
        if boxes is not None:
            text_lines = []
            for box, score in zip(boxes, scores):
                box = sort_poly(box.astype(np.int32))
				
                angle = get_slope(box )
                if math.fabs(angle) > 5:
                    continue
					
                if np.linalg.norm(box[0] - box[1]) < 5 or np.linalg.norm(box[3]-box[0]) < 5:
                    continue
                tl = collections.OrderedDict(zip(
                    ['x0', 'y0', 'x1', 'y1', 'x2', 'y2', 'x3', 'y3'],
                    map(float, box.flatten())))
                tl['score'] = float(score)
                text_lines.append(tl)
        ret = {
            'text_lines': text_lines,
            'rtparams': rtparams,
            'timing': timer,
        }
        #ret.update(get_host_info())
        return ret


    return predictor


### the webserver
from flask import Flask, request, render_template
import argparse


class Config:
    SAVE_DIR = 'static/results'


config = Config()


app = Flask(__name__)

@app.route('/')
def index():
    return render_template('index.html', session_id='dummy_session_id')


def draw_illu(illu, rst):
    for t in rst['text_lines']:
        d = np.array([t['x0'], t['y0'], t['x1'], t['y1'], t['x2'],
                      t['y2'], t['x3'], t['y3']], dtype='int32')
        d = d.reshape(-1, 2)
        cv2.polylines(illu, [d], isClosed=True, color=(255, 255, 0))
    return illu


def save_result(img, rst):
    session_id = str(uuid.uuid1())
    dirpath = os.path.join(config.SAVE_DIR, session_id)
    os.makedirs(dirpath)

    # save input image
    output_path = os.path.join(dirpath, 'input.png')
    cv2.imwrite(output_path, img)

    # save illustration
    output_path = os.path.join(dirpath, 'output.png')
    cv2.imwrite(output_path, draw_illu(img.copy(), rst))

    # save json data
    output_path = os.path.join(dirpath, 'result.json')
    with open(output_path, 'w') as f:
        json.dump(rst, f)

    rst['session_id'] = session_id
    return rst



checkpoint_path = './east_icdar2015_resnet_v1_50_rbox'


@app.route('/', methods=['POST'])
def index_post():
    global predictor
    import io
    bio = io.BytesIO()
    request.files['image'].save(bio)
    img = cv2.imdecode(np.frombuffer(bio.getvalue(), dtype='uint8'), 1)
    rst = get_predictor(checkpoint_path)(img)

    save_result(img, rst)
    return render_template('index.html', session_id=rst['session_id'])


def main():
    global checkpoint_path
    parser = argparse.ArgumentParser()
    parser.add_argument('--port', default=8769, type=int)
    parser.add_argument('--checkpoint_path', default=checkpoint_path)
    parser.add_argument('--debug', action='store_true')
    args = parser.parse_args()
    checkpoint_path = args.checkpoint_path

    if not os.path.exists(args.checkpoint_path):
        raise RuntimeError(
            'Checkpoint `{}` not found'.format(args.checkpoint_path))

    app.debug = args.debug
    app.run('0.0.0.0', args.port)

if __name__ == '__main__':
    main()

**Comment by zqm840527** [comment]
[run_demo_server.txt](https://github.com/argman/EAST/files/2776405/run_demo_server.txt)

LOL not familiar with github, don't know how to upload .py file...

Any one who need this just change the .txt to .py after you download. 

BTW, don't forget to backup your original run_demo_server.py.
